streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBPrefixIterator.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.rocksdb.RocksIterator;
+
+import java.util.Comparator;
+import java.util.Set;
+
+class RocksDBPrefixIterator extends RocksDbIterator {
+    private byte[] rawPrefix;
+
+    RocksDBPrefixIterator(final String name,
+                          final RocksIterator newIterator,
+                          final Set<KeyValueIterator<Bytes, byte[]>> openIterators,
+                          final Bytes prefix) {
+        super(name, newIterator, openIterators);
+        rawPrefix = prefix.get();
+        newIterator.seek(rawPrefix);
+    }
+
+    @Override
+    public synchronized boolean hasNext() {
+        if (!super.hasNext()) {
+            return false;
+        }
+
+        final byte[] rawNextKey = super.peekNextKey().get();
```

Comment:
Just a thought, does Rocks return in sorted order? If so, we can probably
do better by comparing backwards from the end of the prefix.
Then again, that might screw up cache locality, not sure...

Reply:
According to their wiki ( https://github.com/facebook/rocksdb/wiki/Basic-Operations ) it does provide a sorted iterator. The range function operates on this principle too, which confirms this to be the case in the code.

We have to validate the entire prefix for each event either way, forwards or backwards. As soon as we run into an event which does not have the full prefix, we terminate and hasNext() will from then on return false.

I don't think there's anything in terms of efficiency to be done here, but let me know if I misunderstood.